src/imitation/algorithms/bc.py

```diff
@@ -81,54 +43,79 @@ def __init__(
         pairs.
 
         Args:
-            env: environment to train on.
+            observation_space: the observation space of the environment.
+            action_space: the action space of the environment.
             policy_class: used to instantiate imitation policy.
+            policy_kwargs: keyword arguments passed to policy's constructor.
             expert_data: If not None, then immediately call
                   `self.set_expert_dataset(expert_data)` during initialization.
             batch_size: batch size used for training.
             optimizer_cls: optimiser to use for supervised training.
-            optimizer_kwargs: keyword arguments for optimiser construction.
+            optimizer_kwargs: keyword arguments, excluding learning rate and
+                  weight decay, for optimiser construction.
+            ent_weight: scaling applied to the policy's entropy regularization.
+            l2_weight: scaling applied to the policy's L2 regularization.
         """
+        if optimizer_kwargs:
+            if "lr" in optimizer_kwargs:
+                raise ValueError(
+                    "The policy's lr_schedule determines the learning rate."
```

Comment:
Forcing the policy to be `ActorCriticPolicy` seems like a big downside -- no obvious reason we should need this for BC. Seems relatively easy to refactor to accept `BasePolicy` -- the user will just have to be careful to provide a compatible set of optimizer and policy kwargs.

I'm not sure why SB3 has an optimizer in the policy, @araffin would know. My guess is it's because we might want different optimizers for policies and critics.

Reply:
It seems to me, then, that the best thing to do is refactor Behavioral Clone so that it instantiates its own learning rate and optimizer rather than using those of the underlying policy. This refactoring will allow relaxing the policy requirement to a `BasePolicy`, and it should also make it more transparent to save / load the Behavioral Clone because the Behavioral Clone class will store all the relevant attributes (rather than reaching for them inside the policy).

Do we want to move away from the `policy_class` / `policy_kwargs` argument scheme now? Or should I just do the refactoring with a `policy_class` / `policy_kwargs` argument scheme, and we can change this scheme later if we want?

Reply:
Okay, I did this refactoring! `BC` now creates its own optimizer.